watch-library/hardware/watch/watch_private_cdc.c

```diff
@@ -0,0 +1,129 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "watch_private_cdc.h"
+
+#include "watch_utility.h"
+
+#include "tusb.h"
+
+/*
+ * Basic write buffer for USB CDC serial.
+ */
+
+#define CDC_WRITE_BUF_SZ  (256)
+static char s_write_buf[CDC_WRITE_BUF_SZ] = {0};
+static size_t s_write_buf_len = 0;
+
+/*
+ * Implement a circular buffer for the USB CDC Serial read buffer.
+ * The size of the buffer must be a power of two for this circular buffer
+ * implementation to work.
+ */
+
+// Size of the circular buffer. Must be a power of two.
+#define CDC_READ_BUF_SZ  (256)
+// Macro function to perform modular arithmetic on an index.
+// eg. (63 + 2) & (64 - 1) -> 1
+#define CDC_READ_BUF_IDX(x)  ((x) & (CDC_READ_BUF_SZ - 1))
+static char s_read_buf[CDC_READ_BUF_SZ] = {0};
+static size_t s_read_buf_pos = 0;
+static size_t s_read_buf_len = 0;
+
+int _write(int file, char *ptr, int len) {
+    (void) file;
+    if (ptr == NULL || len == 0) {
+        return -1;
+    }
+
+    int32_t bytes_free = CDC_WRITE_BUF_SZ - ((int32_t) s_write_buf_len);
+    if (bytes_free <= 0) {
+        return 0;
+    }
+
+    size_t to_write = min(len, bytes_free);
+    memcpy(&s_write_buf[s_write_buf_len], ptr, to_write);
+    s_write_buf_len += to_write;
+    return to_write;
+}
+
+int _read(int file, char *ptr, int len) {
+    (void)file;
+
+    if (ptr == NULL || len <= 0 || s_read_buf_len == 0) {
+        return -1;
+    }
+
+    // Clamp to the length of the read buffer
+    if ((size_t) len > s_read_buf_len) {
+        len = s_read_buf_len;
+    }
+
+    // Calculate the start of the circular buffer, and iterate from there
+    const size_t start_pos = CDC_READ_BUF_IDX(s_read_buf_pos - len);
+    for (size_t i = 0; i < (size_t) len; i++) {
+        const size_t idx = CDC_READ_BUF_IDX(start_pos + i);
+        ptr[i] = s_read_buf[idx];
+        s_read_buf[idx] = 0;
+    }
+
+    // Update circular buffer position and length
+    s_read_buf_len -= len;
+    s_read_buf_pos = CDC_READ_BUF_IDX(s_read_buf_pos + len);
+
+    return len;
+}
+
+void cdc_task(void) {
+    // Handle Writes
+    if (s_write_buf_len > 0) {
+        int32_t to_write = s_write_buf_len;
+        size_t written = 0;
+        // Write in chunks of 32 bytes at most
```

Comment:
This is the sort of comment that I like having a 'because' on.

It seems to me you're trying to fight against the code in tud_cdc_write, which automatically flushes at > 64 bytes (i.e. theoretically you don't need the intermediate tud_cdc_write_flushes here if this was reasonable).

Should we just be patching tinyusb?

Reply:
You caught me using weasel words in the comments. :) To be honest, I bashed out the USB tweaks months ago while trying to get prints working, and didn't leave any comments for my future self. It's entirely possible that this change isn't needed, but it's been stable for me, so I figured I should just get what I have into a PR for now. From what I recall, it was quite sensitive to the TC0 timer period + how much data we try to place in the FIFO, and required a lot of guess-and-checking of various parameters.

Reply:
Ok, I tried hooking `_write()` directly up to `tud_cdc_write()`, and it more or less seems to work. Just need to make sure `CFG_TUD_CDC_TX_BUFSIZE` is large enough, and need to call `tud_cdc_write_flush()`, or the data won't actually be sent until several characters are written. I'll play around with it and push an updated patch.